File: backend/src/lib/geofence.ts
interface Coordinates {
  lat: number;
  lng: number;
}

const EARTH_RADIUS_METERS = 6371000;

/**
 * The check-in fence around every outlet, in metres. POST /visits/checkin
 * rejects anything further out; the fraud engine reasons in the same fence
 * (stock_outside_outlet, #248), so both read it from here.
 */
export const GEOFENCE_RADIUS_M = 50;

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function haversineDistanceMeters(a: Coordinates, b: Coordinates): number {
  const dLat = toRadians(b.lat - a.lat);
  const dLng = toRadians(b.lng - a.lng);
  const lat1 = toRadians(a.lat);
  const lat2 = toRadians(b.lat);

  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;

  return 2 * EARTH_RADIUS_METERS * Math.asin(Math.sqrt(h));
}

export function isWithinGeofence(
  outlet: Coordinates,
  checkin: Coordinates,
  radiusMeters = GEOFENCE_RADIUS_M,
): boolean {
  return haversineDistanceMeters(outlet, checkin) <= radiusMeters;
}

File: app/lib/core/geo/geofence.dart
import 'dart:math';

class Coordinates {
  const Coordinates({required this.lat, required this.lng});
  final double lat;
  final double lng;
}

const _earthRadiusMeters = 6371000.0;

const defaultGeofenceRadiusMeters = 50.0;

double _toRadians(double degrees) => degrees * pi / 180;

double haversineDistanceMeters(Coordinates a, Coordinates b) {
  final dLat = _toRadians(b.lat - a.lat);
  final dLng = _toRadians(b.lng - a.lng);
  final lat1 = _toRadians(a.lat);
  final lat2 = _toRadians(b.lat);

  final h =
      pow(sin(dLat / 2), 2) + cos(lat1) * cos(lat2) * pow(sin(dLng / 2), 2);

  return 2 * _earthRadiusMeters * asin(sqrt(h));
}

bool isWithinGeofence(
  Coordinates outlet,
  Coordinates checkin, {
  double radiusMeters = defaultGeofenceRadiusMeters,
}) {
  return haversineDistanceMeters(outlet, checkin) <= radiusMeters;
}
